core/blockchain.go

```diff
@@ -2605,3 +2605,29 @@ func (bc *BlockChain) SetTrieFlushInterval(interval time.Duration) {
 func (bc *BlockChain) GetTrieFlushInterval() time.Duration {
 	return time.Duration(bc.flushInterval.Load())
 }
+
+// VerifyBlock checks block state
+func (bc *BlockChain) VerifyBlock(block *types.Block) error {
+	err := bc.validator.ValidateBody(block)
+	if err != nil {
+		return err
+	}
+
+	parent := bc.GetBlockByHash(block.ParentHash())
+	statedb, err := bc.StateAt(parent.Root())
+	if err != nil {
+		err = fmt.Errorf("failed to retrieve state at %s: %w", parent.Root(), err)
+		return err
+	}
+
+	receipts, _, usedGas, err := bc.processor.Process(block, statedb, bc.vmConfig)
+	if err != nil {
+		err = fmt.Errorf("failed to process block at %s: %w", block.Hash(), err)
+		return err
```

Comment:
```suggestion
		return fmt.Errorf("failed to process block: %w", err)
```

Do not call `block.Hash()` here, the block is not finalized (no witnesses).